File: src/main/java/banane/io/pdb/model/Hero.java
package banane.io.pdb.model;

import com.fasterxml.jackson.annotation.JsonGetter;
import com.fasterxml.jackson.annotation.JsonIgnore;

import javax.persistence.*;

@Entity
public class Hero {

    private Long id;

    private String username;

    @JsonIgnore
    private User owner;

    private MapPoint currentZone;

    private Integer strength;

    private Integer agility;

    private Integer intelligence;

    private Integer hp;

    private Integer mana;

    private Integer wood;

    private Integer stone;

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }


    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id",unique = true)
    public User getOwner() {
        return owner;
    }

    public void setOwner(User owner) {
        this.owner = owner;
    }

    public String getUsername() {
        return username;
    }

    public void setUsername(String username) {
        this.username = username;
    }

    @ManyToOne(fetch = FetchType.EAGER, optional = false)
    @JoinColumn(name = "map_point_id", nullable = false)
    public MapPoint getCurrentZone() {
        return currentZone;
    }

    public void setCurrentZone(MapPoint currentZone) {
        this.currentZone = currentZone;
    }

    @Transient
    @JsonGetter("currentZone")
    public Long getCurrentZoneJson() {
        return currentZone.getId();
    }

    public Integer getStrength() {
        return strength;
    }

    public void setStrength(Integer strength) {
        this.strength = strength;
    }

    public Integer getAgility() {
        return agility;
    }

    public void setAgility(Integer agility) {
        this.agility = agility;
    }

    public Integer getIntelligence() {
        return intelligence;
    }

    public void setIntelligence(Integer intelligence) {
        this.intelligence = intelligence;
    }

    public Integer getHp() {
        return hp;
    }

    public void setHp(Integer hp) {
        this.hp = hp;
    }

    public Integer getMana() {
        return mana;
    }

    public void setMana(Integer mana) {
        this.mana = mana;
    }

    public Integer getWood() {
        return wood;
    }

    public void setWood(Integer wood) {
        this.wood = wood;
    }

    public Integer getStone() {
        return stone;
    }

    public void setStone(Integer stone) {
        this.stone = stone;
    }
}

File: src/main/kotlin/banane/io/pdb/model/Hero.kt
package banane.io.pdb.model

import com.fasterxml.jackson.annotation.JsonGetter
import com.fasterxml.jackson.annotation.JsonIgnore
import javax.persistence.*

@Entity
data class Hero(
        @get:GeneratedValue(strategy = GenerationType.IDENTITY)
        @get:Id
        var id: Long? = null,
        var username: String? = null,
        @get:JoinColumn(name = "user_id", unique = true)
        @get:OneToOne(fetch = FetchType.LAZY)
        @JsonIgnore
        var owner: User? = null,
        @get:JoinColumn(name = "map_point_id", nullable = false)
        @get:ManyToOne(fetch = FetchType.EAGER, optional = false)
        var currentZone: MapPoint? = null,
        var strength: Int? = null,
        var agility: Int? = null,
        var intelligence: Int? = null,
        var hp: Int? = null,
        var mana: Int? = null,
        var wood: Int? = null,
        var stone: Int? = null,
        @get:OneToOne(mappedBy = "owner", cascade = [CascadeType.ALL], fetch = FetchType.LAZY)
        @JsonIgnore
        var base: Base? = null
) {
    @get:JsonGetter("currentZone")
    @get:Transient
    val currentZoneJson: Long?
        get() = currentZone?.id

}